### PR TITLE
change #none to none for use-waypoint

### LINKED ISSUE #50841: reserve none for use-waypoint
reserve `none` instead of `#none` for use-waypoint label

### PR DESCRIPTION
**Pull request by ilrudie**
**Please provide a description of this PR:**
fixes #50841 

changes `#none` to `none` for use-waypoint
removes the totally undocumented `~` optional alias for opting out of a waypoint
updates original release note to include correct information about how this feature will function

### CODE REVIEW

**Comment by ilrudie** [comment]
cc @linsun and @louiscryan

File: pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go
Comment (bleggett) [inline pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go:149]:
```suggestion
                // NOTE yes, this means waypoints named `none` will be ignored.
                // Don't name your waypoints that.
		if labelValue == "none" {
```

**Review by linsun — APPROVED** [review]
LGTM but i don't have approval for this component

**Review by bleggett — APPROVED** [review]

**Review by louiscryan — APPROVED** [review]
LGTM

**Review by keithmattix — COMMENTED** [review]
Have we updated the doc for this? Curious why this was changed

**Comment by ilrudie** [comment]
@keithmattix, The `#` prefix is not used elsewhere making this a bit awkward. In `waypoint-for` a traffic type of `none` (without the hash) can be chosen for instance. It was noted while working on documentation and a few folks noted that the # seemed like a typo.

**Review by MorrisLaw — APPROVED** [review]

**Review by keithmattix — APPROVED** [review]

File: pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go
Comment (ilrudie) [reply pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go:151]: seemed a bit snarky to me but I did try a revised wording with, I think, similarly helpful information

File: pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go:151]: How can we make sure `none` is not a valid waypoint name

File: pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go
Comment (costinm) [inline pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go:150]: To be honest I liked the idea behind "#none" and "-"  

I expect we'll need other values to indicate non-Istio gateways or gateways in a different system. The same applies to the 'data plane mode'.

Not against this PR - we don't use "#" in other places - but I suggest adopting the naming using by Sidecar.egress and VirtualServices.gateways, with ~ to indicate no sidecar ( and in the future NAMESPACE/NAME to indicate a waypoint in a different namespace - and other extensions based on "path label" and presence of "/" separator ). 

"~" is used in Sidecar and I think we should continue to support it here as well ( even if we keep none too ).

Comment (ilrudie) [reply pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go:150]: actually this brings up an interesting point. A user likely could have a waypoint called `none` if we allow `namespace/name` format but they will not be able to use the short form. They could use `my-cool-namespace/none` to get around the limitation of reserving `none`.

That being said something like `~` isn't a bad shout either and I'm happy to put it back in. I had removed it because it wasn't documented but I'm not hard opposed to it.

Comment (linsun) [reply pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go:150]: Not sure if you saw discussion on #ambient-dev @costinm, a few folks thought #none was a typo in our doc.  Given it is not consistent with anywhere else in Istio, we don't see why we just reserve `none` as a reserved name.

supporting `~` is okay if you feel very strongly about it but I don't want to doc it given it is not documented by sidecars.

Comment (ilrudie) [reply pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go:150]: I did add back ~ and I also went ahead and excluded a suffix of "/none" to prevent attempts to work around the reserved wording by the example I had provided above with namespace/none

File: pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go
Comment (ilrudie) [reply pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go]: https://github.com/istio/istio/pull/50842 was created to implements preventing istioctl users from choosing the reserved word `none` for their waypoint's name

**Comment by costinm** [comment]
On Mon, May 6, 2024 at 8:47 AM Lin Sun ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go
> <https://github.com/istio/istio/pull/50842#discussion_r1591200571>:
>
> > @@ -146,7 +146,9 @@ func fetchWaypointForWorkload(ctx krt.HandlerContext, Waypoints krt.Collection[W
>  // defaultNamespace avoids the need to infer when object meta from a namespace was given
>  func getUseWaypoint(meta metav1.ObjectMeta, defaultNamespace string) (named *krt.Named, isNone bool) {
>  	if labelValue, ok := meta.Labels[constants.AmbientUseWaypointLabel]; ok {
> -		if labelValue == "#none" || labelValue == "~" {
> +		// NOTE: this means Istio reserves the word "none" in this field with a special meaning
> +		//   a waypoint named "none" cannot be used and will be ignored
>
> Not sure if you saw discussion on #ambient-dev @costinm
> <https://github.com/costinm>, a few folks thought #none was a typo in our
> doc. Given it is not consistent with anywhere else in Istio, we don't see
> why we just reserve none as a reserved name.
>
Confused.

I agree "#none" is not consistent and not the best option.

I don't have strong opinions about using "none" as a reserved name - we do
use it in a few places ( interception mode at least ).

supporting ~ is okay if you feel very strongly about it but I don't want to
> doc it given it is not documented by sidecars.
>

"~" is documented in the Sidecar API, and doesn't have some of the problems
that a reserved name would have ( having a namespace or service named
"none" is maybe
low chance, but still not zero ). We also use names containing "/" - in
VirtualService "gateways" for example.

I wouldn't say I feel very strongly about "~", but I like consistency.

**Comment by linsun** [comment]
> On Mon, May 6, 2024 at 8:47 AM Lin Sun ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go <[#50842 (comment)](https://github.com/istio/istio/pull/50842#discussion_r1591200571)>: > @@ -146,7 +146,9 @@ func fetchWaypointForWorkload(ctx krt.HandlerContext, Waypoints krt.Collection[W // defaultNamespace avoids the need to infer when object meta from a namespace was given func getUseWaypoint(meta metav1.ObjectMeta, defaultNamespace string) (named *krt.Named, isNone bool) { if labelValue, ok := meta.Labels[constants.AmbientUseWaypointLabel]; ok { - if labelValue == "#none" || labelValue == "~" { + // NOTE: this means Istio reserves the word "none" in this field with a special meaning + // a waypoint named "none" cannot be used and will be ignored Not sure if you saw discussion on #ambient-dev @costinm <https://github.com/costinm>, a few folks thought #none was a typo in our doc. Given it is not consistent with anywhere else in Istio, we don't see why we just reserve none as a reserved name.
> Confused. I agree "#none" is not consistent and not the best option. I don't have strong opinions about using "none" as a reserved name - we do use it in a few places ( interception mode at least ). supporting ~ is okay if you feel very strongly about it but I don't want to
> doc it given it is not documented by sidecars.
> "~" is documented in the Sidecar API, and doesn't have some of the problems that a reserved name would have ( having a namespace or service named "none" is maybe low chance, but still not zero ). We also use names containing "/" - in VirtualService "gateways" for example. I wouldn't say I feel very strongly about "~", but I like consistency. Message ID: ***@***.***>
> […](#)

Ok I saw `~` is documented in Sidecar resource (as a value for hosts) which I missed earlier. Given the value of hosts are `*` or `.`, it makes sense to use `~` there.  What I meant is we don't support `istio.io/injection=~` afaik.

**Comment by costinm** [comment]
On Mon, May 6, 2024 at 12:10 PM Lin Sun ***@***.***> wrote:

> On Mon, May 6, 2024 at 8:47 AM Lin Sun *@*.*> wrote: @.** commented on
> this pull request. ------------------------------ In
> pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go <#50842
> (comment)
> <https://github.com/istio/istio/pull/50842#discussion_r1591200571>>: > @@
> -146,7 +146,9 @@ func fetchWaypointForWorkload(ctx krt.HandlerContext,
> Waypoints krt.Collection[W // defaultNamespace avoids the need to infer
> when object meta from a namespace was given func getUseWaypoint(meta
> metav1.ObjectMeta, defaultNamespace string) (named
>
>
> *krt.Named, isNone bool) { if labelValue, ok :=
> meta.Labels[constants.AmbientUseWaypointLabel]; ok { - if labelValue ==
> "#none" || labelValue == "" { + // NOTE: this means Istio reserves the word
> "none" in this field with a special meaning + // a waypoint named "none"
> cannot be used and will be ignored Not sure if you saw discussion on
> #ambient-dev @costinm <https://github.com/costinm>
> https://github.com/costinm <https://github.com/costinm>, a few folks
> thought #none was a typo in our doc. Given it is not consistent with
> anywhere else in Istio, we don't see why we just reserve none as a reserved
> name. Confused. I agree "#none" is not consistent and not the best option.
> I don't have strong opinions about using "none" as a reserved name - we do
> use it in a few places ( interception mode at least ). supporting ~ is okay
> if you feel very strongly about it but I don't want to doc it given it is
> not documented by sidecars. "" is documented in the Sidecar API, and
> doesn't have some of the problems that a reserved name would have ( having
> a namespace or service named "none" is maybe low chance, but still not zero
> ). We also use names containing "/" - in VirtualService "gateways" for
> example. I wouldn't say I feel very strongly about "~", but I like
> consistency. Message ID: @.***>
> … <#m_-5179901243563530929_>
>
> Ok I saw ~ is documented in Sidecar resource (as a value for hosts) which
> I missed earlier. Given the value of hosts are * or ., it makes sense to
> use ~ there. What I meant is we don't support istio.io/injection=~ afaik.
>

Sure - we don't support injecton=none either.

"~" wouldn't be my first choice - I didn't like it when it was added in
Sidecar - but better than inventing yet another 'magic word' in a label.

The semantics of VirtualService.gateways and Sidecar.egress are closer to
what we are doing here - on defining config scoping/delegation.

File: pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go
Comment (craigbox) [reply pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go:150]: I'm up for documenting it in both places

**Comment by ilrudie** [comment]
@istio/wg-networking-maintainers-pilot, is anyone able to review and provide feedback on whether this is suitable to move forward or requires further refinements?

**Review by howardjohn — APPROVED** [review]

**Pull request merged**
